### PR TITLE
Validate metadata hash

### LINKED ISSUE #73: Validate MDS data
For now, we are ignoring the hash validation (obtained from the TOC) of the downloaded MDS statement.

We should probably validate it.

### PR DESCRIPTION
**Pull request by tpazderka**
Close #73

### CODE REVIEW

File: django_fido/management/commands/download_authenticator_metadata.py [hunk 1]
```diff
@@ -1,22 +1,77 @@
...
+
+def verify_certificate(jwt: JWT) -> JWK:
+    """Get (and verify) the signing key from JWT."""
+    # First element in the header is our actual key
+    try:
+        decoding_key = JWK.from_pem(PEM_CERT_TEMPLATE.format(jwt.token.jose_header['x5c'][0]).encode())
+    except ValueError:
+        raise InvalidCert('Cannot decode key.')
+    if SETTINGS.metadata_service['certificate'] is None:
+        return decoding_key
```
Comment (ziima) [inline django_fido/management/commands/download_authenticator_metadata.py:45]: Disabling verification should be done the same way as in metadata.

File: django_fido/management/commands/download_authenticator_metadata.py [hunk 2]
```diff
@@ -1,22 +1,77 @@
...
+    # Create crypto context
+    store = crypto.X509Store()
+    for key in jwt.token.jose_header['x5c'][1:]:
+        cert = crypto.load_certificate(crypto.FILETYPE_PEM, PEM_CERT_TEMPLATE.format(key).encode())
+        store.add_cert(cert)
+    with open(str(SETTINGS.metadata_service['certificate'])) as root_file:
```
Comment (ziima) [inline django_fido/management/commands/download_authenticator_metadata.py:51]: Why `str`?

File: django_fido/management/commands/download_authenticator_metadata.py [hunk 3]
```diff
@@ -57,7 +113,12 @@ def handle(self, **options):
         except TypeError:
             raise CommandError('access_token setting must be specified for this command to work.')
 
-        metadata = _get_metadata()
+        metadata, alg = _get_metadata()
+        # Convert alg name to corresponding hashing algorithm
+        if alg[0:2].upper() in ('HS', 'RS', 'ES'):
+            alg = 'SHA' + alg[2:]
```
Comment (ziima) [inline django_fido/management/commands/download_authenticator_metadata.py:119]: I'd rather have a mapping for algoritms.

File: django_fido/management/commands/download_authenticator_metadata.py [hunk 4]
```diff
@@ -73,5 +134,9 @@ def handle(self, **options):
             authenticator.metadata_entry = json.dumps(authenticator_data)
             auth_metadata = requests.get(url, params={'token': SETTINGS.metadata_service['access_token']},
                                          timeout=SETTINGS.metadata_service['timeout'])
-            authenticator.detailed_metadata_entry = urlsafe_b64decode(auth_metadata.content).decode()
+            hash = hashlib.new(alg, auth_metadata.content)
```
Comment (ziima) [inline django_fido/management/commands/download_authenticator_metadata.py:137]: I believe supported algorithm may vary.

File: django_fido/management/commands/download_authenticator_metadata.py (hunk 1 above)
Comment (tpazderka) [reply django_fido/management/commands/download_authenticator_metadata.py:45]: This is from #70, will go away after a rebase.

File: django_fido/management/commands/download_authenticator_metadata.py (hunk 2 above)
Comment (tpazderka) [reply django_fido/management/commands/download_authenticator_metadata.py:51]: Python 3.5 cannot handle `Path` object which is a result of `FileSetting`.

File: django_fido/management/commands/download_authenticator_metadata.py (hunk 4 above)
Comment (tpazderka) [reply django_fido/management/commands/download_authenticator_metadata.py:137]: Only SHA-256, SHA-384 and SHA-512 are specified in JWA specifications and they are in the guaranteed alg for hashlib, so they are always available.

File: django_fido/management/commands/download_authenticator_metadata.py [hunk 5]
```diff
@@ -23,6 +24,16 @@ class InvalidCert(Exception):
...
 
 
+def urlsafe_b64decode(decodable: bytes) -> bytes:
+    """Alternative implementation to fill the necessary padding."""
+    m = len(decodable) % 4
+    if m == 2:
+        decodable += b'=='
+    elif m == 3:
+        decodable += b'='
+    return base64.urlsafe_b64decode(decodable)
```
Comment (stinovlas) [inline django_fido/management/commands/download_authenticator_metadata.py:34]: Shouldn't there be a padding added for `m == 1` as well?

File: django_fido/management/commands/download_authenticator_metadata.py (hunk 3 above)
Comment (stinovlas) [reply django_fido/management/commands/download_authenticator_metadata.py:129]: I agree with the mapping.

File: django_fido/management/commands/download_authenticator_metadata.py (hunk 5 above)
Comment (tpazderka) [reply django_fido/management/commands/download_authenticator_metadata.py:34]: That is an invalid remainder and the string is not base64 encoded. https://tools.ietf.org/html/rfc4648#section-4

**Comment by tpazderka** [comment]
Mapping added.

File: django_fido/management/commands/download_authenticator_metadata.py (hunk 2 above)
Comment (tpazderka) [reply django_fido/management/commands/download_authenticator_metadata.py]: And this is also from #70

File: django_fido/management/commands/download_authenticator_metadata.py (hunk 4 above)
Comment (tpazderka) [reply django_fido/management/commands/download_authenticator_metadata.py]: Check for supported algs added anyway.

File: django_fido/management/commands/download_authenticator_metadata.py (hunk 5 above)
Comment (stinovlas) [reply django_fido/management/commands/download_authenticator_metadata.py:34]: Thanks for the clarification, you're right.

File: django_fido/management/commands/download_authenticator_metadata.py (hunk 3 above)
Comment (stinovlas) [reply django_fido/management/commands/download_authenticator_metadata.py]: `ALG_MAPPING` is not very descriptive. What about `HASH_ALG_MAPPING`?

**Pull request merged**